src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,401 @@
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
+        }
+
+        if (request.path.mDataVersion.Value() != clusterInfo->dataVersion)
+        {
+            ChipLogError(DataManagement, "Write Version mismatch for Endpoint 0x%x, Cluster " ChipLogFormatMEI,
+                         request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
+        }
+    }
+
+    AttributeAccessInterface * aai       = GetAttributeAccessOverride(request.path.mEndpointId, request.path.mClusterId);
+    std::optional<CHIP_ERROR> aai_result = TryWriteViaAccessInterface(request.path, aai, decoder);
+    if (aai_result.has_value())
+    {
+        if (*aai_result == CHIP_NO_ERROR)
+        {
+            // TODO: change callbacks should likely be routed through the context `MarkDirty` only
+            //       however for now this is called directly because ember code does this call
+            //       inside emberAfWriteAttribute.
+            MatterReportingAttributeChangeCallback(request.path);
+            CurrentContext().dataModelChangeListener->MarkDirty(request.path);
+        }
+        return *aai_result;
+    }
+
+    MutableByteSpan dataBuffer = gEmberAttributeIOBufferSpan;
+    ReturnErrorOnFailure(DecodeValueIntoEmberBuffer(decoder, *attributeMetadata, dataBuffer));
+
+    Protocols::InteractionModel::Status status;
+
+    if (request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        // Internal requests use the non-External interface that has less enforcement
+        // than the external version (e.g. does not check/enforce writable settings, does not
+        // validate attribute types) - see attribute-table.h documentation for details.
+        status = emberAfWriteAttribute(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId,
+                                       gEmberAttributeIOBufferSpan.data(), (*attributeMetadata)->attributeType);
+    }
+    else
+    {
+        if (dataBuffer.size() > (*attributeMetadata)->size)
+        {
+            ChipLogDetail(Zcl, "Data to write exceeds the attribute size claimed.");
+            return CHIP_IM_GLOBAL_STATUS(InvalidValue);
+        }
+
+        status = emAfWriteAttributeExternal(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId,
+                                            gEmberAttributeIOBufferSpan.data(), (*attributeMetadata)->attributeType);
+    }
+
+    if (status != Protocols::InteractionModel::Status::Success)
+    {
+        return CHIP_ERROR_IM_GLOBAL_STATUS_VALUE(status);
+    }
+
+    // TODO: this may need more refinement:
+    //       - should internal requests be able to decide if something is marked dirty or not?
+    //       - changes-omitted paths should not be marked dirty (ember is not aware of these)
+    CurrentContext().dataModelChangeListener->MarkDirty(request.path);
```

Comment:
This looks wrong.  `emberAfWriteAttribute` will not mark things dirty if the value is not actually changing.  At least in some cases.

Followup, not for this PR, since the MarkDirty() stuff is currently all a no-op, I assume.  But this needs some thought about what the API will _actually_ look like, and the comment should be very clear that this _definitely_ needs "refinement".

Reply:
Re-worded comment and added details about what to consider regarding MatterReportingAttributeChangeCallback.

It seems a bit annoying that AAI always calls it (regardless of content) where as ember has knowledge of changed or not. This needs more iteration to make it consistent.

Reply:
@andy31415 Yes, we probably need to change the AAI API to allow the impl to say "actually, nothing changed" for impls that want to optimize.